### ISSUE TITLE
Negative indices skip increasing order requirement

### ISSUE THREAD

**Issue by shoyer** [h5py/h5py#993]
```
import h5py
import numpy as np

f = h5py.File('/tmp/test.h5', 'w')
f.create_dataset('data', data=np.arange(3))
f['data'][[2, 1]]
```
This raises:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/conda/envs/xarray-py36/lib/python3.6/site-packages/h5py/_hl/selections.py in select(shape, args, dsid)
     84             try:
---> 85                 int(a)
     86                 if isinstance(a, np.ndarray) and a.shape == (1,):

TypeError: int() argument must be a string, a bytes-like object or a number, not 'list'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-18-3a0bcbaae636> in <module>()
----> 1 f['data'][[2, 1]]

h5py/_objects.pyx in h5py._objects.with_phil.wrapper (/Users/ilan/minonda/conda-bld/h5py_1496887972496/work/h5py/_objects.c:2846)()

h5py/_objects.pyx in h5py._objects.with_phil.wrapper (/Users/ilan/minonda/conda-bld/h5py_1496887972496/work/h5py/_objects.c:2804)()

~/conda/envs/xarray-py36/lib/python3.6/site-packages/h5py/_hl/dataset.py in __getitem__(self, args)
    472
    473         # Perform the dataspace selection.
--> 474         selection = sel.select(self.shape, args, dsid=self.id)
    475
    476         if selection.nselect == 0:

~/conda/envs/xarray-py36/lib/python3.6/site-packages/h5py/_hl/selections.py in select(shape, args, dsid)
     88             except Exception:
     89                 sel = FancySelection(shape)
---> 90                 sel[args]
     91                 return sel
     92

~/conda/envs/xarray-py36/lib/python3.6/site-packages/h5py/_hl/selections.py in __getitem__(self, args)
    359                     adjacent = zip(list_arg[:-1], list_arg[1:])
    360                     if any(fst >= snd for fst, snd in adjacent):
--> 361                         raise TypeError("Indexing elements must be in increasing order")
    362
    363         if len(sequenceargs) > 1:

TypeError: Indexing elements must be in increasing order
```

However, I can convert this requirement by using -1 to index from the end instead:
```
f['data'][[-1, 1]]
# array([1, 2])
```

Summary of the h5py configuration
---------------------------------

h5py    2.7.0
HDF5    1.8.17
Python  3.6.3 | packaged by conda-forge | (default, Oct  5 2017, 19:18:17)
[GCC 4.2.1 Compatible Apple LLVM 6.1.0 (clang-602.0.53)]
sys.platform    darwin
sys.maxsize     9223372036854775807
numpy   1.13.1

**Comment by aragilar** [comment]
I'm guessing this should be prevented (and maybe look at why there are the current restrictions on type and order, as to whether they can be relaxed or not).

**Comment by takluyver** [comment]
Dropping the milestone for now because no-one appears to be working on it.

**Comment by takluyver** [comment]
Fixed in 3.0 - it now checks the order after wrapping negative indices to positive ones.
